onnxruntime/core/providers/shared_library/provider_author.h

```diff
@@ -0,0 +1,281 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// Provider authors include this file
+
+#pragma once
+
+#include <vector>
+#include <string>
+#include <map>
+#include <unordered_map>
+#include "onnx/common/stl_backports.h"
+#include "core/common/common.h"
+#include "core/common/const_pointer_container.h"
+#include "core/session/onnxruntime_c_api.h"
+#include "provider_interfaces.h"
+
+namespace ONNX_NAMESPACE {
+
+enum AttributeProto_AttributeType : int {
+  AttributeProto_AttributeType_UNDEFINED = 0,
+  AttributeProto_AttributeType_FLOAT = 1,
+  AttributeProto_AttributeType_INT = 2,
+  AttributeProto_AttributeType_STRING = 3,
+  AttributeProto_AttributeType_TENSOR = 4,
+  AttributeProto_AttributeType_GRAPH = 5,
+  AttributeProto_AttributeType_SPARSE_TENSOR = 11,
+  AttributeProto_AttributeType_FLOATS = 6,
+  AttributeProto_AttributeType_INTS = 7,
+  AttributeProto_AttributeType_STRINGS = 8,
+  AttributeProto_AttributeType_TENSORS = 9,
+  AttributeProto_AttributeType_GRAPHS = 10,
+  AttributeProto_AttributeType_SPARSE_TENSORS = 12
+};
+
+enum OperatorStatus : int {
+  EXPERIMENTAL = 0,
+  STABLE = 1
+};
+
+}  // namespace ONNX_NAMESPACE
+
+namespace onnxruntime {
+
+// The function passed in will be run on provider DLL unload. This is used to free thread_local variables that are in threads we don't own
+// Since these are not destroyed when the DLL unloads we have to do it manually. Search for usage for an example.
+struct RunOnUnload {
+  RunOnUnload(std::function<void()> run);
+  ~RunOnUnload();
+
+ private:
+  bool* enabled_{};
+};
+
+constexpr const char* kOnnxDomain = "";
+constexpr const char* kDnnlExecutionProvider = "DnnlExecutionProvider";
+
+class DataTypeImpl {
+ public:
+  virtual ~DataTypeImpl() = default;
+
+  template <typename T>
+  static MLDataType GetType();
+  template <typename elemT>
+  static MLDataType GetTensorType();
+};
+
+class TensorShape : private std::vector<int64_t> {
+ public:
+  TensorShape() = default;
+
+  TensorShape(const TensorShape& /*other*/) = default;
+  TensorShape& operator=(const TensorShape& /*other*/) = default;
+
+  TensorShape(TensorShape&& /*other*/) = default;
+  TensorShape& operator=(TensorShape&& /*other*/) = default;
+
+  TensorShape(const std::vector<int64_t>& dims) : std::vector<int64_t>{dims} {}
+  TensorShape(std::vector<int64_t>&& dims) : std::vector<int64_t>{dims} {}
+  TensorShape(const std::initializer_list<int64_t>& dims) : std::vector<int64_t>{dims} {}
+
+  TensorShape(const int64_t* dimension_sizes, size_t dimension_count);
+  TensorShape(const std::vector<int64_t>& dims, size_t start, size_t end);
+
+  using std::vector<int64_t>::operator[];
+
+#if 0
```

Comment:
Do we need this? Same for many other places.

Reply:
I removed them all. I kept needing to enable more and more but at this point it's not needed.